### PR TITLE
Amesos2: Implement Umfpack wrappers for Amesos2

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
The current status:
This will run the solver_test for Umfpack on Amesos2 and
pass in serial for the Epetra and Tpetra tests.

This is in progress and this PR is for discussion purposes.

Note that the current state won't compile with complex.
That is in progress and needs some discussion.

Need to resolve some warnings with SuiteSparse included.

Need to implement Transpose.

### CODE REVIEW

**Comment by MicheldeMessieres** [comment]
@srajama1 I set up this PR for the current state of the Umfpack wrappers for Amesos2 for our discussion meeting. Complex still needs to be set up (will not compile in current state) and some discussion there will probably help quite a bit. In Serial solver_test will pass for the Epetra and Tpetra tests and the wrappers are modeled on the SuperLu wrappers.

**Comment by mhoemmen** [comment]
@vbrunini this might be of interest to you

**Comment by MicheldeMessieres** [comment]
@srajama1 This PR is ready for review. Note that linking to an externally built Suite Sparse (required for Umfpack) will give warnings due to name space collisions. Those warnings are resolved by a separate PR #1202 as we discussed.

**Comment by MicheldeMessieres** [comment]
@srajama1 Let me know what you'd prefer for the CMake warnings for double/float. If complex will never be allowed for float only I can just eliminate that one. It sounds like this will be a very low priority but perhaps leaving them in doesn't cause any harm.

**Comment by srajama1** [comment]
@jwillenbring : I am reviewing these changes. Considering this is an user request it would be nice if we start nightly testing with UMFPACK. Can the @trilinos/framework team add such a nightly one to run the test @MicheldeMessieres has added. This can be a separate issue for the framework and I can go ahead and accept these changes first.

File: packages/amesos2/CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +20,15 @@ IF (NOT Tpetra_INST_INT_INT)
 
 ENDIF ()
 
+IF (Amesos2_ENABLE_UMFPACK)
+  IF (Trilinos_ENABLE_COMPLEX_FLOAT AND NOT Trilinos_ENABLE_COMPLEX_DOUBLE)
+    MESSAGE(WARNING "*****Amesos2 is building Umfpack without a complex<double> type. Umfpack natively supports only complex<double> types, not complex<float> types, and this will lead to type conversions. *****")
+  ENDIF()
+  # This may not happen anyways… right now it seems we can’t compile this package with Trilinos_ENABLE_FLOAT and TPetra_INST_DOUBLE turned off - need to resolve.
+  IF (NOT Tpetra_INST_DOUBLE)
```
Comment (srajama1) [inline packages/amesos2/CMakeLists.txt]: Can we use a Trilinos level Macro ?

File: packages/amesos2/src/Amesos2_Umfpack.cpp [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
```
Comment (srajama1) [inline packages/amesos2/src/Amesos2_Umfpack.cpp]: Can we change this to my name and e-mail.

File: packages/amesos2/src/Amesos2_Umfpack_FunctionMap.hpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+      extern void
+      umfpack_di_defaults(double Control [UMFPACK_CONTROL]);
+      extern void
+      umfpack_di_free_numeric(void **);
+      extern void
+      umfpack_di_free_symbolic(void **);
+    }
+
+#ifdef HAVE_TEUCHOS_COMPLEX
+    namespace Z {              // double-precision complex definitions
```
Comment (srajama1) [inline packages/amesos2/src/Amesos2_Umfpack_FunctionMap.hpp]: Technically this namespace is not needed

File: packages/amesos2/CMakeLists.txt (hunk 1 above)
Comment (krcb) [inline packages/amesos2/CMakeLists.txt]: There does not appear to be a Trillions level equivalent of the Tpetra_INST_DOUBLE macro.

File: packages/amesos2/src/Amesos2_Umfpack_TypeMap.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+
+#include "Amesos2_Umfpack_TypeMap.hpp"
+
```
Comment (srajama1) [inline packages/amesos2/src/Amesos2_Umfpack_TypeMap.cpp]: Is this empty file needed ?

File: packages/amesos2/src/Amesos2_Umfpack_TypeMap.hpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+#include "Amesos2_TypeMap.hpp"
+
+/* Umfpack comples headers file only need to be included if
+   complex has been enabled in Teuchos.  In addition we only need to
+   define the conversion and printing functions if complex has been
+   enabled. */
+namespace UMFPACK {
+
+// Declare and specialize a std::binary_funtion class for
+// multiplication of Umfpack types
```
Comment (srajama1) [inline packages/amesos2/src/Amesos2_Umfpack_TypeMap.hpp]: Should this be in FunctionMap ?

File: packages/amesos2/src/Amesos2_Umfpack_decl.hpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  Teuchos::Array<int> colptr_;
+
+  /// Persisting 1D store for X
+  Teuchos::Array<umfpack_type> xvals_;  int ldx_;
+  /// Persisting 1D store for B
+  Teuchos::Array<umfpack_type> bvals_;  int ldb_;
+};                              // End class Umfpack
+
+
+// Specialize solver_traits struct for Umfpack
```
Comment (srajama1) [inline packages/amesos2/src/Amesos2_Umfpack_decl.hpp]: TBD : Check

**Comment by bartlettroscoe** [comment]
> > Can we use a Trilinos level Macro ?
>
>There does not appear to be a Trillinos level equivalent of the Tpetra_INST_DOUBLE macro.

Mark and I had a long documented discussion about this in https://github.com/trilinos/Trilinos/issues/362#issuecomment-261989166.  The consensus is that there is very little value in allowing users to enable or disable double support.  Therefore, if explicit instantiation support is enabled, then instantiations should be provided for 'double', period. If this does not seem logical, then we should discuss this at the next Trilinos Leaders Meeting and if it can't wait for that, then we should start a project-wide discussion on this topic.

**Comment by srajama1** [comment]
Well this is not as big as #362. This is a portion of that. We just wanted a macro that allow us to check if double is ON or OFF. The macro could always be ON for now. The trouble was there was no such macro, so we ended up using the Tpetra macro reluctantly.

**Comment by bartlettroscoe** [comment]
> We just wanted a macro that allow us to check if double is ON or OFF. The macro could always be ON for now. The trouble was there was no such macro, so we ended up using the Tpetra macro reluctantly.

Right.  Unless someone can make a compelling argument for disabling double support, then we should just assume that it is always enabled.  But if you just want a macro to tell you if Amesos2 should provide double instantiations, then that is fine.  That is easy to do in the Amesos2 CMakeLists.txt and Amesos2_config.h.in files.  That should be a two line change (one in each file).

**Comment by srajama1** [comment]
Isn't the what Tpetra did ? Instead of defining this macro multiple packages, we went looking for a Trilinos level macro. Is that a small change as well ?

**Comment by bartlettroscoe** [comment]
> Isn't the what Tpetra did ? Instead of defining this macro multiple packages, we went looking for a Trilinos level macro. Is that a small change as well ?

What I am saying is that a user should never see an option to enable or disable `double` support.  The only question is if to create explicit instantations for `double`.  And for that, your Trilinos package should just check `${PACKAGE_NAME}_ENABLE_EXPLICIT_TEMPLATE_INSTANTIATION` as per:

* https://trilinos.org/docs/files/TrilinosBuildReference.html#enabling-explicit-template-instantiation-for-c

You package can define and use **namespaced** macros based on `${PACKAGE_NAME}_ENABLE_EXPLICIT_TEMPLATE_INSTANTIATION` any way it likes.

Given that ETI can be enabled and disabled on a package-by-package basis, you should not be checking what Tpetra does with ETI,  You should just be checking `${PACKAGE_NAME}_ENABLE_EXPLICIT_TEMPLATE_INSTANTIATION`.

Does that make sense?  If not, I can update this PR with a commit that does the right thing to demonstrate.

**Comment by srajama1** [comment]
This is not for explicit instantiation in a package, but to check double is tuned on or off, and generate a warning based on that.

**Comment by MicheldeMessieres** [comment]
@srajama1 Updated based on our discussion. Changed e-mail names to you, removed the D and Z name spacing, removed the unused umfpack_multi, removed the empty TypeDef.cpp file. I confirmed that the templated make_list code we discussed is necessary. Did not update the CMakeLists warnings yet - please let me know if you want that changed.

**Comment by srajama1** [comment]
This will get pushed in after the Suitesparse namespace changes in Trilinos. This also requires nightly testing framework issue #1255.

**Comment by MicheldeMessieres** [comment]
This was tested on SEMS and passed. Will apply merge.

**Pull request merged**